submissions/BlueLamantine/task-js-opp/app.js

```diff
@@ -0,0 +1,102 @@
+const WIDTH = 505;
+const BLOCK_HEIGHT = 77;
+const STEP = 90;
+const MAX_SPEED = 400;
+const MIN_SPEED = 100;
+const PLAYER_CONF = {
+    startPosX : 200, 
+    startPosY : 400,
+};
+
+const ENEMY_CONF = {
+    quantity : 3,
+    startPosX : -100,
+    startPosY : [1,2,3].map(rowNumber => rowNumber * BLOCK_HEIGHT),
+};
+
+let Enemy = function(x,y, speed) {
+    this.x = x;
+    this.y = y;
+    this.speed = speed;
+    this.sprite = 'images/cat.gif';
+};
+
+let allEnemies =  Array.from(Array(ENEMY_CONF.quantity),
+(val, index) => new Enemy(ENEMY_CONF.startPosX,ENEMY_CONF.startPosY[index],getRandomNumber(MIN_SPEED,MAX_SPEED)));
+
+Enemy.prototype.update = function(dt) {
+    this.x = this.x > WIDTH ? ENEMY_CONF.startPosX : this.x + this.speed * dt;
+};
+
+Enemy.prototype.render = function() {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+let Player = function() {
+    this.x = PLAYER_CONF.startPosX;
+    this.y = PLAYER_CONF.startPosY;
+    this.sprite = 'images/fish.gif';
+};
+Player.prototype.setDefault = function () {
+    this.x = PLAYER_CONF.startPosX;
+    this.y = PLAYER_CONF.startPosY;
+};
+
+Player.prototype.update = function() {
+    if (this.y < 0) {
+        alert('WINNER, you saved the fish!');
+        player.setDefault();
+    }
+    this.checkCollision();
+};
+Player.prototype.checkCollision = function () {
+    allEnemies.forEach(function(enemy) {
+        if ( Math.round(enemy.x / STEP) === Math.round(player.x / STEP) && 
+             Math.round(enemy.y / BLOCK_HEIGHT) === Math.round(player.y / BLOCK_HEIGHT) ) {
+                alert('you lose! =(');
+                player.setDefault();
```

Comment:
It is against OOP and SOLID principles to refer to global variables from inside the class.

Exclusions are constants and public API.

`allEnemies` and `player` are global variables here.

`allEnemies`: just let `Player` instance legally know about them at a player construction time.

`player`: we do need to refer to a player variable here at all as we are already "inside" the player. Hope it is obvious how to refer the object from inside itself.

Reply:
Ok, I fixed . The function ```checkCollision();``` does not refer to global variables except constants (hope that present global variable ```allEnemies``` as a parameter of constructor ```Player``` is a legal way)

Reply:
It's absolutely legal. This way a entity (character) learns improtant things about surrounding world.